File: grid/internal/webaccount/privileges.go
package webaccount

import (
	"errors"
	"regexp"
	"sort"
	"strings"
)

// Named administrative privileges. See api/README.md "Privileges".
const (
	PrivUsers    = "users"
	PrivBans     = "bans"
	PrivRegions  = "regions"
	PrivMap      = "map"
	PrivDeploy   = "deploy"
	PrivUndeploy = "undeploy"
	PrivAdmin    = "admin"
	PrivSuper    = "super"
)

// adminExpansion is the deliberately curated common set that `admin` grants.
// The destructive `undeploy` capability is intentionally excluded and must be
// assigned separately.
var adminExpansion = []string{PrivUsers, PrivBans, PrivRegions, PrivMap, PrivDeploy}

// allNamedPrivileges is every privilege `super` grants; it is also the set of
// privileges an endpoint can require.
var allNamedPrivileges = []string{
	PrivUsers, PrivBans, PrivRegions, PrivMap, PrivDeploy, PrivUndeploy, PrivAdmin, PrivSuper,
}

// ErrInvalidPrivileges is returned when a privilege string violates the
// Privileges schema pattern.
var ErrInvalidPrivileges = errors.New("privileges must be a normalized comma-separated list of names")

var privilegeName = regexp.MustCompile(`^[a-z][a-z0-9_-]*$`)

// NormalizePrivileges validates each name against the schema pattern and
// returns a deduplicated, sorted, comma-separated string. An empty input is
// valid and yields an empty string.
func NormalizePrivileges(privs string) (string, error) {
	trimmed := strings.TrimSpace(privs)
	if trimmed == "" {
		return "", nil
	}
	seen := map[string]struct{}{}
	names := make([]string, 0)
	for _, name := range strings.Split(trimmed, ",") {
		if !privilegeName.MatchString(name) {
			return "", ErrInvalidPrivileges
		}
		if _, exists := seen[name]; exists {
			continue
		}
		seen[name] = struct{}{}
		names = append(names, name)
	}
	sort.Strings(names)
	return strings.Join(names, ","), nil
}

// expand resolves a stored privilege string into the concrete set of granted
// capabilities, applying the admin and super expansions. Unknown names confer
// no authority.
func expand(privs string) map[string]bool {
	granted := map[string]bool{}
	for _, name := range strings.Split(privs, ",") {
		name = strings.TrimSpace(name)
		if name == "" {
			continue
		}
		granted[name] = true
		switch name {
		case PrivSuper:
			for _, p := range allNamedPrivileges {
				granted[p] = true
			}
		case PrivAdmin:
			for _, p := range adminExpansion {
				granted[p] = true
			}
		}
	}
	return granted
}

// HasPrivilege reports whether the stored privilege string authorizes the
// required capability, honoring admin/super expansion.
func HasPrivilege(privs, required string) bool {
	if required == "" {
		return true
	}
	return expand(privs)[required]
}

// IsSuper reports whether the stored privilege string includes super.
func IsSuper(privs string) bool {
	return expand(privs)[PrivSuper]
}

File: web/src/lib/privileges.js
// Mirrors the server privilege model (grid/internal/webaccount/privileges.go).
// The API is authoritative; this mirror gates the admin UI so it only offers
// actions the account can actually perform. Keep the two in sync.

// Named administrative privileges, in a stable display order.
export const NAMED_PRIVILEGES = [
  "users",
  "bans",
  "regions",
  "map",
  "deploy",
  "undeploy",
  "system",
  "admin",
  "super",
];

// `admin` expands to this curated common set. The destructive `undeploy`
// capability is intentionally excluded and must be assigned separately.
const ADMIN_EXPANSION = ["users", "bans", "regions", "map", "deploy"];

// Splits a stored privilege string into its raw, stored names.
export function privilegeList(privs) {
  return String(privs ?? "")
    .split(",")
    .map((name) => name.trim())
    .filter(Boolean);
}

// Resolves a stored privilege string into the concrete set of granted
// capabilities, applying the admin and super expansions. Unknown names confer
// no authority.
function expand(privs) {
  const granted = new Set();
  for (const name of privilegeList(privs)) {
    granted.add(name);
    if (name === "super") {
      for (const priv of NAMED_PRIVILEGES) {
        granted.add(priv);
      }
    } else if (name === "admin") {
      for (const priv of ADMIN_EXPANSION) {
        granted.add(priv);
      }
    }
  }
  return granted;
}

// Reports whether the stored privilege string authorizes the required
// capability, honoring admin/super expansion. An empty requirement is always
// authorized.
export function hasPrivilege(privs, required) {
  if (!required) {
    return true;
  }
  return expand(privs).has(required);
}

// Reports whether the stored privilege string includes super.
export function isSuper(privs) {
  return expand(privs).has("super");
}

// The privileges that grant access to some part of the admin area.
const ADMIN_SURFACE = ["users", "bans", "regions", "map", "deploy", "undeploy", "system"];

// Reports whether the account may reach any admin area, used to decide whether
// to surface the Admin navigation entry at all.
export function hasAnyAdmin(privs) {
  const granted = expand(privs);
  return ADMIN_SURFACE.some((priv) => granted.has(priv));
}

// The full, expanded set of effective privileges as a sorted array, for
// display alongside the stored set.
export function effectivePrivileges(privs) {
  return [...expand(privs)].sort();
}
